File: leetcode/ruby/isValidBST.rb
# Given a binary tree, determine if it is a valid binary search tree (BST).

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}


def is_valid_bst(root, min = nil, max = nil) 
    return true if !root
    return false if min && root.val <= min
    return false if max && root.val >= max
    return is_valid_bst(root.left, min, root.val) && is_valid_bst(root.right, root.val, max)
end

File: leetcode/js/isValidBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root, min, max) {
  if (!root) return true;
  if (min !== undefined && root.val <= min) {
    return false;
  }
  if (max !== undefined && root.val >= max) {
    return false;
  }
  return (
    isValidBST(root.left, min, root.val) &&
    isValidBST(root.right, root.val, max)
  );
};
